src/main/java/org/springframework/data/neo4j/core/mapping/DefaultNeo4jEntityConverter.java

```diff
@@ -476,40 +492,36 @@ private static Value extractValueOf(Neo4jPersistentProperty property, MapAccesso
 
 		private final Map<Long, Object> internalIdStore = new HashMap<>();
 
-		Object computeIfAbsent(Long internalId, Supplier<Object> entitySupplier) {
-			Object knownEntity = getObject(internalId);
-
-			// if it is not in the store, it has to get re-computed also for the internalIdStore
-			if (knownEntity != null) {
-				return knownEntity;
+		private void storeObject(@Nullable Long internalId, Object object) {
+			if (internalId == null) {
+				return;
 			}
-
 			try {
 				write.lock();
-				Object computedEntity = entitySupplier.get();
-				internalIdStore.put(internalId, computedEntity);
-				return computedEntity;
+				internalIdStore.put(internalId, object);
 			} finally {
 				write.unlock();
 			}
 		}
 
-		@Nullable
-		private Object getObject(Long internalId) {
+		private Optional<Object> getObject(@Nullable Long internalId) {
+			if (internalId == null) {
+				return Optional.empty();
+			}
 			try {
 
 				read.lock();
 
 				Object knownEntity = internalIdStore.get(internalId);
 
 				if (knownEntity != null) {
-					return knownEntity;
+					return Optional.of(knownEntity);
```

Comment:
Can we do without the optionals here?

Reply:
For references: Exactly this change was done in the follow up PR #541